### PR TITLE
module-4

### PR DESCRIPTION
**Pull request by AliakseiSayarkin**

### CODE REVIEW

File: src/main/java/com/epam/esm/security/UserSecurity.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        }
+    }
+
+    public boolean hasUserId(Authentication authentication, EntityModel<OrderDto> order) {
+        try {
+            UserDetailsImpl details = (UserDetailsImpl) authentication.getPrincipal();
+
+            return Objects.requireNonNull(order.getContent()).getUserId() == details.getId();
+        } catch (ClassCastException e) {
+            return false;
```
Comment (churaalena) [inline src/main/java/com/epam/esm/security/UserSecurity.java]: Why not to throw some exception?

File: src/main/java/com/epam/esm/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -95,4 +96,17 @@ public int getLastPage(int size) throws ServiceException {
...
     }
+
+    @Override
+    @Transactional(rollbackFor = ServiceException.class)
+    public User addUser(User user) throws ServiceException {
+        userValidator.validateUser(user);
+        try {
+            return userDao.addUser(user);
+        } catch (DataAccessException | NoResultException | IllegalArgumentException e) {
+            e.printStackTrace();
```
Comment (churaalena) [inline src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: Why do we need stacktrace here?

File: src/main/java/com/epam/esm/web/controller/AuthController.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    public void logout(HttpServletRequest request, HttpServletResponse response) {
+        SecurityContextLogoutHandler securityContextLogoutHandler = new SecurityContextLogoutHandler();
+        securityContextLogoutHandler.logout(request, response, null);
+    }
+
+    private Map<String, Object> getResponse(AuthRequestDto requestDto, User user) {
+        String token = jwtTokenProvider.createJwtToken(
+                requestDto.getLogin(), user.getRole().getRoleType().toString()
+        );
+        Map<String, Object> response = new HashMap<>();
```
Comment (churaalena) [inline src/main/java/com/epam/esm/web/controller/AuthController.java]: We can create an object

File: src/main/java/com/epam/esm/service/impl/UserServiceImpl.java (hunk 2 above)
Comment (AliakseiSayarkin) [reply src/main/java/com/epam/esm/service/impl/UserServiceImpl.java]: removed stack trace

File: src/main/java/com/epam/esm/web/controller/AuthController.java (hunk 3 above)
Comment (AliakseiSayarkin) [reply src/main/java/com/epam/esm/web/controller/AuthController.java]: Added object

File: src/main/java/com/epam/esm/security/UserSecurityUtil.java [hunk 4]
```diff
@@ -24,12 +24,13 @@ public boolean hasUserId(Authentication authentication, int userId) {
...
+    public boolean authenticateOrderId(Authentication authentication, int orderId) {
         try {
             var details = (UserDetailsImpl) authentication.getPrincipal();
             var order = auditedOrderService.getAuditedOrderById(orderId);
+            var user = order.getUser();
 
-            return order.getUser().getId() == details.getId();
+            return user != null && user.getId() == details.getId();
         } catch (ClassCastException | ServiceException e) {
             return false;
```
Comment (churaalena) [inline src/main/java/com/epam/esm/security/UserSecurityUtil.java:41]: Where is logging?

File: src/test/java/com/epam/esm/security/UserSecurityUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.security.authentication.AuthenticationServiceException;
+import org.springframework.security.core.userdetails.UserDetailsService;
+
+@SpringBootTest
```
Comment (churaalena) [inline src/test/java/com/epam/esm/security/UserSecurityUtilTest.java]: Why do you need this annotation for unit tests?

File: src/test/java/com/epam/esm/security/UserSecurityUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        var username = "username";
+        var role = "USER";
+        var token = jwtTokenProvider.createJwtToken(username, role);
+
+        Mockito.when(userDetailsService.loadUserByUsername(username))
+                .thenReturn(new UserDetailsImpl(1, username, "password", null, true));
+
+        var auth = jwtTokenProvider.getAuthentication(token);
+        var userDetails = (UserDetailsImpl) auth.getPrincipal();
+        Assertions.assertNotNull(auth);
```
Comment (churaalena) [inline src/test/java/com/epam/esm/security/UserSecurityUtilTest.java]: Please use static import to increase readability

File: src/main/java/com/epam/esm/security/UserSecurityUtil.java (hunk 4 above)
Comment (AliakseiSayarkin) [reply src/main/java/com/epam/esm/security/UserSecurityUtil.java:41]: Added logging

File: src/test/java/com/epam/esm/security/UserSecurityUtilTest.java (hunk 5 above)
Comment (AliakseiSayarkin) [reply src/test/java/com/epam/esm/security/UserSecurityUtilTest.java]: To properly mock objects

File: src/test/java/com/epam/esm/security/UserSecurityUtilTest.java (hunk 6 above)
Comment (AliakseiSayarkin) [reply src/test/java/com/epam/esm/security/UserSecurityUtilTest.java]: Done

**Pull request merged**